compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirSuperReferenceChecker.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.KtNodeTypes.TYPE_ARGUMENT_LIST
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getChild
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.references.FirSuperReference
+import org.jetbrains.kotlin.fir.types.ConeKotlinErrorType
+import org.jetbrains.kotlin.fir.types.coneType
+import org.jetbrains.kotlin.psi.stubs.elements.KtStubElementTypes.TYPE_ELEMENT_TYPES
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+object FirSuperReferenceChecker : FirQualifiedAccessExpressionChecker() {
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        val superReference = expression.calleeReference.safeAs<FirSuperReference>()?.takeIf { it.hadExplicitTypeInSource() } ?: return
+
+        val typeArgumentListSource = superReference.superTypeRef.source?.getChild(TYPE_ELEMENT_TYPES)?.getChild(TYPE_ARGUMENT_LIST)
```

Comment:
In this case, we have to navigate to find the type argument because FIR tree does not store whether such explicit type references are present or not. We can of course still report the super expression and thenuse a position strategy to locate the same type reference to report. But this seems unnecessarily complicated.

Reply:
Hi, can you elaborate how we can do this without calling `getChild`? IIUC, the FIR tree does not tell us if the type argument is explicitly set or not.

Reply:
Please take my changes from https://github.com/JetBrains/kotlin/commit/43dcc9c7a8b2824e9784dec8021e5ec5fd29457b (you can squash my commit with your changes).

Explanation: we store additional information that can be used in checkers or on IDE side to `delegatedTypeRef`. This includes nested `typeRef`, `source` elements. For more details please take a look at the function [`extractArgumentTypeRefAndSource`](https://github.com/JetBrains/kotlin/blob/173813f7cf829059b4f7b15aa0aedfcc0b7818a2/compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/FirHelpers.kt#L600) and its usages.

Reply:
Hi! Any chance to update this respective to my suggestion?

Reply:
Sorry I totally missed your early comment! I have cherry picked your commit on top of mine and everything seems to work fine. Thanks!